api/app/models/env_canada.py

```diff
@@ -169,48 +174,105 @@ def flag_file_as_processed(session, url):
-                    # extract model info from filename:
-                    model_info = parse_env_canada_filename(filename)
-                    # download the file:
-                    downloaded = download(url, gdps_path)
-                    if downloaded:
-                        files_downloaded += 1
-                        # If we've downloaded the file ok, we can now process it.
-                        try:
-                            processor.process_grib_file(downloaded, model_info)
-                            # Flag the file as processed
-                            flag_file_as_processed(session, url)
-                            files_processed += 1
-                        finally:
-                            # delete the file when done.
-                            os.remove(downloaded)
-            # pylint: disable=broad-except
-            except Exception as exception:
-                exception_count += 1
-                # We catch and log exceptions, but keep trying to download.
-                # We intentionally catch a broad exception, as we want to try and download as much
-                # as we can.
-                logger.error('unexpected exception processing %s',
-                             url, exc_info=exception)
+    # We always work in UTC:
+    now = get_utcnow()
+    for hour in get_model_run_hours():
+        logger.info('Processing GDPS model run {:02d}'.format(hour))
+        urls = get_model_run_download_urls(now, hour)
+        processor = GribFileProcessor()
+        with tempfile.TemporaryDirectory() as tmp_path:
+            for url, filename in urls:
+                try:
+                    # check the database for a record of this file:
+                    processed_file_record = get_processed_file_record(
+                        session, url)
+                    if processed_file_record:
+                        # This file has already been processed - so we skip it.
+                        logger.info('file aready processed %s', url)
+                    else:
+                        # extract model info from filename:
+                        model_info = parse_env_canada_filename(filename)
+                        # download the file:
+                        downloaded = download(url, tmp_path)
+                        if downloaded:
+                            files_downloaded += 1
+                            # If we've downloaded the file ok, we can now process it.
+                            try:
+                                processor.process_grib_file(
+                                    downloaded, model_info)
+                                # Flag the file as processed
+                                flag_file_as_processed(session, url)
+                                files_processed += 1
+                            finally:
+                                # delete the file when done.
+                                os.remove(downloaded)
```

Comment:
Do we need another try block inside?

Reply:
That try is dealing specifically with cleanup of the file, in that context. So yeah - I'd argue we want it, even though it's ugly. Should we chat? Do you have a more elegant proposal?

Reply:
Sonarcloud doesn't like the cognitive complexity in this section of code though - so I'm going to re-factor it a bit.